### PR TITLE
add aave v3 decoder

### PR DESCRIPTION
**Pull request by akbaridria**
Add AAVE V3 Decoder that includes:
- borrow
- supply
- withdraw
- repay
- flashloan
- liquidationcall

### CODE REVIEW

File: services/decoder/protocols/aave-v3/aave-v3.decoders.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import WithdrawDecoder from "./pools/withdraw";
+import FlashloanDecoder from "./pools/flashloan";
+import LiquidationDecoder from "./pools/liquidation";
+
+BorrowDecoder;
+SupplyDecoder;
+RepayDecoder;
+WithdrawDecoder;
+FlashloanDecoder;
+LiquidationDecoder;
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/aave-v3.decoders.ts:13]: Dont create a separate folder `pools` and import the decoding functions from there, rather keep all the decoding functions in `aave-v3.decoders.ts`

File: services/decoder/protocols/aave-v3/pools/borrow.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        }) as {
+            eventName: "Borrow";
+            args: {
+                reserve: string;
+                user: string;
+                onBehalfOf: string;
+                amount: bigint;
+                interestRateMode: INTEREST_RATE_MODE;
+                borrowRate: bigint;
+                referalCode: bigint;
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/borrow.ts:42]: It is not `referalCode`, but `referralCode` - that's why its throwing undefined

File: services/decoder/protocols/aave-v3/pools/borrow.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+export const BorrowDecoder = GoldRushDecoder.on(
+    "aave-v3:Borrow",
+    ["eth-mainnet"],
+    ABI as Abi,
+    async (log_event, tx, chain_name, covalent_client): Promise<EventType> => {
+        const { raw_log_data, raw_log_topics } = log_event;
+
+        enum INTEREST_RATE_MODE {
+            "STABLE" = 1,
+            "VARIABLE" = 2,
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/borrow.ts:25]: Fix casing of the keys to sentence case (Stable, Variable)

File: services/decoder/protocols/aave-v3/pools/borrow.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            },
+            {
+                title: "On Behalf Of",
+                value: decoded.onBehalfOf,
+                type: "address",
+            },
+            {
+                title: "Borrow Rate",
+                value: String(decoded.borrowRate),
+                type: "text",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/borrow.ts:65]: The Borrow Rate seems like gibberish - try to calculate what its actual value stands for and parse it accordingly

File: services/decoder/protocols/aave-v3/pools/borrow.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                decoded.reserve,
+                {
+                    from: date,
+                    to: date,
+                }
+            );
+
+        const tokens: EventTokens = [
+            {
+                decimals: BorrowToken?.[0].contract_decimals,
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/borrow.ts:94]: Optional `?` chain this - `BorrowToken?.[0]?.contract_decimals`

File: services/decoder/protocols/aave-v3/pools/borrow.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                    BorrowToken?.[0]?.prices?.[0].price *
+                        (Number(decoded.amount) /
+                            Math.pow(
+                                10,
+                                BorrowToken?.[0]?.contract_decimals ?? 0
+                            ))
+                ),
+                ticker_logo: BorrowToken?.[0].logo_url,
+                ticker_symbol: BorrowToken?.[0].contract_ticker_symbol,
+                value: String(decoded.amount),
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/borrow.ts:106]: Optional `?` chain

File: services/decoder/protocols/aave-v3/pools/borrow.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            },
+        ];
+
+        return {
+            action: DECODED_ACTION.TRANSFERRED,
+            category: DECODED_EVENT_CATEGORY.LENDING,
+            name: "Borrow",
+            protocol: {
+                logo: log_event.sender_logo_url as string,
+                name: "AAVE-V3",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/borrow.ts:116]: The name of the protocol should be `Aave V3`

File: services/decoder/protocols/aave-v3/pools/supply.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            },
+        ];
+
+        return {
+            action: DECODED_ACTION.TRANSFERRED,
+            category: DECODED_EVENT_CATEGORY.LENDING,
+            name: "Supply",
+            protocol: {
+                logo: log_event.sender_logo_url as string,
+                name: "AAVE-V3",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/supply.ts:96]: Fix protocol name

File: services/decoder/protocols/aave-v3/pools/repay.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                type: "address",
+            },
+            {
+                title: "Repayer",
+                value: decoded.repayer,
+                type: "address",
+            },
+            {
+                title: "Use A Tokens",
+                value: String(decoded.useATokens),
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/repay.ts:57]: Parse boolean to be either `Yes` or `No`

File: services/decoder/protocols/aave-v3/pools/repay.ts [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+            },
+        ];
+
+        return {
+            action: DECODED_ACTION.TRANSFERRED,
+            category: DECODED_EVENT_CATEGORY.LENDING,
+            name: "Repay",
+            protocol: {
+                logo: log_event.sender_logo_url as string,
+                name: "AAVE-V3",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/repay.ts:99]: Fix protocol name

File: services/decoder/protocols/aave-v3/pools/withdraw.ts [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+            },
+        ];
+
+        return {
+            action: DECODED_ACTION.TRANSFERRED,
+            category: DECODED_EVENT_CATEGORY.LENDING,
+            name: "Withdraw",
+            protocol: {
+                logo: log_event.sender_logo_url as string,
+                name: "AAVE-V3",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/withdraw.ts:93]: Fix protocol name

File: services/decoder/protocols/aave-v3/pools/flashloan.ts [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+            },
+        ];
+
+        return {
+            action: DECODED_ACTION.TRANSFERRED,
+            category: DECODED_EVENT_CATEGORY.LENDING,
+            name: "FlashLoan",
+            protocol: {
+                logo: log_event.sender_logo_url as string,
+                name: "AAVE-V3",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/flashloan.ts:111]: Fix protocol name

File: services/decoder/protocols/aave-v3/pools/flashloan.ts [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+                ticker_logo: FlashloanToken?.[0].logo_url,
+                ticker_symbol: FlashloanToken?.[0].contract_ticker_symbol,
+                value: String(decoded.amount),
+            },
+        ];
+
+        return {
+            action: DECODED_ACTION.TRANSFERRED,
+            category: DECODED_EVENT_CATEGORY.LENDING,
+            name: "FlashLoan",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/flashloan.ts:108]: It should be `Flash Loan`

File: services/decoder/protocols/aave-v3/pools/flashloan.ts [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+export const FlashloanDecoder = GoldRushDecoder.on(
+    "aave-v3:FlashLoan",
+    ["eth-mainnet"],
+    ABI as Abi,
+    async (log_event, tx, chain_name, covalent_client): Promise<EventType> => {
+        const { raw_log_data, raw_log_topics } = log_event;
+
+        enum INTEREST_RATE_MODE {
+            "STABLE" = 1,
+            "VARIABLE" = 2,
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/flashloan.ts:25]: Interest Rate Mode also has a value of 0 - please handle that case

File: services/decoder/protocols/aave-v3/pools/flashloan.ts [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+            },
+            {
+                title: "Interest Rate Mode",
+                value: INTEREST_RATE_MODE[decoded.interestRateMode],
+                type: "text",
+            },
+            {
+                title: "Premium",
+                value: String(decoded.premium),
+                type: "text",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/flashloan.ts:70]: Parse premium into meaningful value

File: services/decoder/protocols/aave-v3/pools/liquidation.ts [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+                value: decoded.collateralAsset,
+                type: "address",
+            },
+            {
+                title: "Debt Asset",
+                value: decoded.debtAsset,
+                type: "address",
+            },
+            {
+                title: "user",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/liquidation.ts:53]: `User`

File: services/decoder/protocols/aave-v3/pools/liquidation.ts [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+                type: "address",
+            },
+            {
+                title: "Liquidator",
+                value: decoded.liquidator,
+                type: "address",
+            },
+            {
+                title: "Receive A Token",
+                value: String(decoded.receiveAToken),
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/liquidation.ts:64]: Parse boolean to a `Yes` or `No`

File: services/decoder/protocols/aave-v3/pools/liquidation.ts [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+                ticker_logo: debtToken?.[0].logo_url,
+                ticker_symbol: debtToken?.[0].contract_ticker_symbol,
+                value: String(decoded.debtToCover),
+            },
+        ];
+
+        return {
+            action: DECODED_ACTION.TRANSFERRED,
+            category: DECODED_EVENT_CATEGORY.LENDING,
+            name: "LiquidationCall",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/liquidation.ts:123]: `Liquidation Call`

File: services/decoder/protocols/aave-v3/pools/liquidation.ts [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+            protocol: {
+                logo: log_event.sender_logo_url as string,
+                name: "AAVE-V3",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/liquidation.ts:126]: `Aave V3`

File: services/decoder/protocols/aave-v3/pools/liquidation.ts [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+                            Math.pow(
+                                10,
+                                debtToken?.[0]?.contract_decimals ?? 0
+                            ))
+                ),
+                ticker_logo: debtToken?.[0].logo_url,
+                ticker_symbol: debtToken?.[0].contract_ticker_symbol,
+                value: String(decoded.debtToCover),
+            },
+        ];
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/liquidation.ts:118]: Optional `?` chain

File: services/decoder/protocols/aave-v3/pools/withdraw.ts [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+                            Math.pow(
+                                10,
+                                RepayToken?.[0]?.contract_decimals ?? 0
+                            ))
+                ),
+                ticker_logo: RepayToken?.[0].logo_url,
+                ticker_symbol: RepayToken?.[0].contract_ticker_symbol,
+                value: String(decoded.amount),
+            },
+        ];
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/withdraw.ts:85]: Optional `?` chain

File: services/decoder/protocols/aave-v3/pools/supply.ts [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+                            Math.pow(
+                                10,
+                                SupplyToken?.[0]?.contract_decimals ?? 0
+                            ))
+                ),
+                ticker_logo: SupplyToken?.[0].logo_url,
+                ticker_symbol: SupplyToken?.[0].contract_ticker_symbol,
+                value: String(decoded.amount),
+            },
+        ];
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/supply.ts:88]: Optional `?` chain

File: services/decoder/protocols/aave-v3/pools/repay.ts [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+                            Math.pow(
+                                10,
+                                RepayToken?.[0]?.contract_decimals ?? 0
+                            ))
+                ),
+                ticker_logo: RepayToken?.[0].logo_url,
+                ticker_symbol: RepayToken?.[0].contract_ticker_symbol,
+                value: String(decoded.amount),
+            },
+        ];
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/repay.ts:91]: Optional `?` chain

File: services/decoder/protocols/aave-v3/pools/flashloan.ts [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+                            Math.pow(
+                                10,
+                                FlashloanToken?.[0]?.contract_decimals ?? 0
+                            ))
+                ),
+                ticker_logo: FlashloanToken?.[0].logo_url,
+                ticker_symbol: FlashloanToken?.[0].contract_ticker_symbol,
+                value: String(decoded.amount),
+            },
+        ];
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/pools/flashloan.ts:103]: Optional `?` chain

**Review by jagnani73 — CHANGES_REQUESTED** [review]

File: services/decoder/protocols/aave-v3/pools/borrow.ts (hunk 3 above)
Comment (jagnani73) [reply services/decoder/protocols/aave-v3/pools/borrow.ts:25]: Handle the value `0`

File: services/decoder/protocols/aave-v3/pools/flashloan.ts (hunk 14 above)
Comment (jagnani73) [reply services/decoder/protocols/aave-v3/pools/flashloan.ts:25]: Fix casing to sentence case (as above)

**Comment by akbaridria** [comment]
Fixed

[✅] folder structure
[✅] naming protocol and event
[✅] interest rate mode case 0 (None)
[✅] option chain
[✅] borrow rate value (express in ray so i convert it with 1e27)
[✅] premium value

File: services/decoder/protocols/aave-v3/aave-v3.decoders.ts [hunk 25]
```diff
@@ -0,0 +1,604 @@
...
+                        10 ** FlashloanToken?.[0]?.contract_decimals
+                ),
+                type: "text",
+            },
+        ];
+
+        const tokens: EventTokens = [
+            {
+                decimals: FlashloanToken?.[0]?.contract_decimals,
+                heading: "FlashLoan Amount",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/aave-v3.decoders.ts:195]: `Flash Loan Amount`

File: services/decoder/protocols/aave-v3/aave-v3.decoders.ts [hunk 26]
```diff
@@ -0,0 +1,604 @@
...
+                value: INTEREST_RATE_MODE[decoded.interestRateMode],
+                type: "text",
+            },
+            {
+                title: "Premium",
+                value: new Intl.NumberFormat().format(
+                    Number(decoded.premium) /
+                        10 ** FlashloanToken?.[0]?.contract_decimals
+                ),
+                type: "text",
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/aave-v3.decoders.ts:188]: Add a unit for the value as well

File: services/decoder/protocols/aave-v3/aave-v3.decoders.ts [hunk 27]
```diff
@@ -0,0 +1,604 @@
...
+                value: decoded.onBehalfOf,
+                type: "address",
+            },
+            {
+                title: "Borrow Rate",
+                value: new Intl.NumberFormat().format(
+                    Number(decoded.borrowRate) / 1e27
+                ),
+                type: "text",
+            },
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/aave-v3.decoders.ts:65]: Add a unit for the rate

**Review by jagnani73 — CHANGES_REQUESTED** [review]

File: services/decoder/protocols/aave-v3/aave-v3.decoders.ts (hunk 27 above)
Comment (jagnani73) [reply services/decoder/protocols/aave-v3/aave-v3.decoders.ts:65]: If applicable

**Comment by akbaridria** [comment]
Fixed:

[✅] add unit value for premium and borrow rate
[✅] fix event name

File: services/decoder/protocols/aave-v3/aave-v3.decoders.ts [hunk 28]
```diff
@@ -181,18 +182,21 @@ GoldRushDecoder.on(
...
-                ),
+                value:
+                    new Intl.NumberFormat().format(
+                        Number(decoded.premium) /
+                            10 ** FlashloanToken?.[0]?.contract_decimals
+                    ) +
+                    " " +
+                    FlashloanToken?.[0]?.contract_ticker_symbol,
                 type: "text",
             },
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/aave-v3.decoders.ts:193]: The Premium should be in tokens, not details

**Review by jagnani73 — CHANGES_REQUESTED** [review]

File: services/decoder/protocols/aave-v3/aave-v3.decoders.ts (hunk 28 above)
Comment (akbaridria) [reply services/decoder/protocols/aave-v3/aave-v3.decoders.ts]: Your are right! fixed

**Review by jagnani73 — APPROVED** [review]

**Comment by jagnani73** [comment]
@akbaridria Aave exists on multiple chains, can you add configs and test for 2-3 more chains?

File: services/decoder/protocols/aave-v3/aave-v3.configs.ts [hunk 29]
```diff
@@ -0,0 +1,12 @@
+import { type Configs } from "../../decoder.types";
+
+const configs: Configs = [
+    {
+        address: "0x87870bca3f3fd6335c3f4ce8392d69350b4fa4e2",
+        is_factory: false,
+        protocol_name: "aave-v3",
+        network: "eth-mainnet",
+    },
+];
```
Comment (jagnani73) [inline services/decoder/protocols/aave-v3/aave-v3.configs.ts:10]: Add configs fore 2-3 more chains

**Review by jagnani73 — CHANGES_REQUESTED** [review]

File: services/decoder/protocols/aave-v3/aave-v3.configs.ts (hunk 29 above)
Comment (akbaridria) [reply services/decoder/protocols/aave-v3/aave-v3.configs.ts]: just added all supported chain for aave v3 except for gnosis, scroll and fantom.

**Pull request merged**